### PR TITLE
Add post_selection and feed_forward properties to program.

### PR DESCRIPTION
**Pull request by felipeoyarce**
**Context:**
As the related issue #302 states

> there is no _nice way_ of checking whether a circuit has feed-forwarding or post-selection.

**Description of the Change:**

New properties added to `Program` that return True/False for when feed-forwarding and/or post-selection is used. These properties are `post-selection` and `feed-forward`.

**Benefits:**

Improves the developer experience because now you can just you use these new properties instead of having to check each operation in the circuit each time you want to run it.

Sample old code for checking this using try-except:
```python
for c in program.circuit:
    try:
        if c.op.select:
            print("post-selection is used")
    except AttributeError:
        pass

    if c.op.measurement_deps:
        print("feed-forwarding is used")
```

**Possible Drawbacks:**
None

**Related GitHub Issues:**
Issue #302

### CODE REVIEW

**Comment by felipeoyarce** [comment]
Hey @thisac feel free to test these properties, they should work properly :hugs:

File: strawberryfields/program.py [hunk 1]
```diff
@@ -433,6 +433,60 @@ def append(self, op, reg):
...
+
+    @property
+    def post_selection(self):
+        """Return a boolean to indicate if any program's operation uses
+        post-selection or not
+
+        Returns:
+            (bool)
+        """
+
```
Comment (thisac) [inline strawberryfields/program.py:474]:
```suggestion
        Returns:
            bool: whether post-selection is used anywhere in the circuit
        """
```

File: strawberryfields/program.py [hunk 2]
```diff
@@ -433,6 +433,60 @@ def append(self, op, reg):
...
+            op (Operation): quantum operation
+        Returns
+            (bool): boolean indicating if the quantum operation uses feed-forwarding
+                or not
+        """
+
+        return bool(op.measurement_deps)
+
+    @property
+    def post_selection(self):
```
Comment (thisac) [inline strawberryfields/program.py:467]: Maybe change this name to something more descriptive. E.g., `has_post_selection`? 🤔

File: strawberryfields/program.py [hunk 3]
```diff
@@ -433,6 +433,60 @@ def append(self, op, reg):
...
+
+        Returns:
+            (bool)
+        """
+
+        is_select_ops = [self._is_select(c.op) for c in self.circuit]
+        return any(is_select_ops)
+
+    @property
+    def feed_forward(self):
```
Comment (thisac) [inline strawberryfields/program.py:479]: Same here regarding name (`has_feed_forward`).

File: strawberryfields/program.py [hunk 4]
```diff
@@ -433,6 +433,60 @@ def append(self, op, reg):
...
+        """Return a boolean to indicate if any program's operation uses
+        feed-forwarding or not
+
+        Returns:
+            (bool)
+        """
+
```
Comment (thisac) [inline strawberryfields/program.py:486]:
```suggestion
        Returns:
            bool: whether feed-forwarding is used anywhere in the circuit
        """
```

**Review by thisac — CHANGES_REQUESTED** [review]
Looks great @felipeoyarce! Works really well. Thanks for taking the time to address this issue. Just a few comments, but mainly minor ones. 💯

File: strawberryfields/program.py [hunk 5]
```diff
@@ -433,6 +433,60 @@ def append(self, op, reg):
...
+        `measurement_deps` is a non empty set.
+
+        Args:
+            op (Operation): quantum operation
+        Returns
+            (bool): boolean indicating if the quantum operation uses feed-forwarding
+                or not
+        """
+
+        return bool(op.measurement_deps)
```
[the remark this answers is not in the document]
Comment (felipeoyarce) [reply strawberryfields/program.py:460]: Perfect! It makes more sense :smile:

File: strawberryfields/program.py (hunk 2 above)
Comment (felipeoyarce) [reply strawberryfields/program.py:463]: Sure, I just used the same _names_ proposed in the issue #302. I'm going to make the changes :+1:

File: strawberryfields/program.py [hunk 6]
```diff
@@ -433,6 +433,54 @@ def append(self, op, reg):
         self.circuit.append(Command(op, reg))
         return reg
 
+    def _is_select(self, op):
+        """Check if the quantum operation uses post-selection.
+
+        A quantum operation uses post-selection if it has the attribute `select`
+        and this attribute is not None.
```
Comment (thisac) [inline strawberryfields/program.py:440]:
```suggestion
        and this attribute is not ``None``.
```

File: strawberryfields/program.py [hunk 7]
```diff
@@ -433,6 +433,54 @@ def append(self, op, reg):
         self.circuit.append(Command(op, reg))
         return reg
 
+    def _is_select(self, op):
+        """Check if the quantum operation uses post-selection.
+
+        A quantum operation uses post-selection if it has the attribute `select`
```
Comment (thisac) [inline strawberryfields/program.py:439]:
```suggestion
        A quantum operation uses post-selection if it has the attribute ``select``
```

File: strawberryfields/program.py [hunk 8]
```diff
@@ -433,6 +433,54 @@ def append(self, op, reg):
...
+        Returns:
+            bool: whether the quantum operation uses post-selection or not
+        """
+        return hasattr(op, "select") and getattr(op, "select") is not None
+
+    def _is_feed_forward(self, op):
+        """Check if the quantum operation uses feed-forwarding.
+
+        A quantum operation uses feed-forwarding if its attribute
+        `measurement_deps` is a non empty set.
```
Comment (thisac) [inline strawberryfields/program.py:453]:
```suggestion
        ``measurement_deps`` is a non empty set.
```

File: strawberryfields/program.py [hunk 9]
```diff
@@ -433,6 +433,54 @@ def append(self, op, reg):
...
+            op (Operation): quantum operation
+        Returns:
+            bool: whether the quantum operation uses feed-forwarding or not
+        """
+        return len(op.measurement_deps) != 0
+
+    @property
+    def has_post_selection(self):
+        """Return a boolean to indicate if any program's operation uses
+        post-selection or not
```
Comment (thisac) [inline strawberryfields/program.py:465]:
```suggestion
        """Indicate if any operation in the program uses post-selection or not.
```

File: strawberryfields/program.py [hunk 10]
```diff
@@ -433,6 +433,54 @@ def append(self, op, reg):
...
+        Returns:
+            bool: whether post-selection is used anywhere in the circuit
+        """
+        is_select_ops = [self._is_select(c.op) for c in self.circuit]
+        return any(is_select_ops)
+
+    @property
+    def has_feed_forward(self):
+        """Return a boolean to indicate if any program's operation uses
+        feed-forwarding or not
```
Comment (thisac) [inline strawberryfields/program.py:476]:
```suggestion
        """Indicate if any operation in the program uses feed-forwarding or not.
```

**Review by thisac — CHANGES_REQUESTED** [review]
Thanks @felipeoyarce. It seems like a lot of the tests have changed slightly (perhaps due to running black on the tests files). It would be good to refrain from changing these here and only focus on the functionality you've added.

Otherwise it looks good, apart from a few changes in some docstrings, e.g. code being represented by double backticks (``) instead of single (`) ones due to us using Sphinx for the docs. 🙂

**Comment by felipeoyarce** [comment]
> Thanks @felipeoyarce. It seems like a lot of the tests have changed slightly (perhaps due to running black on the tests files). It would be good to refrain from changing these here and only focus on the functionality you've added.
> 
> Otherwise it looks good, apart from a few changes in some docstrings, e.g. code being represented by double backticks (``) instead of single (`) ones due to us using Sphinx for the docs.

Thanks @thisac!

**Review by thisac — APPROVED** [review]
Great work @felipeoyarce! Thanks for working on this.

**Comment by josh146** [comment]
Hi @felipeoyarce, thanks for this PR! We have a SF feature freeze at the moment, but will be able to merge this PR in early next week :)

**Comment by felipeoyarce** [comment]
Sure! Thanks @josh146 and @thisac :hugs:

**Comment by co9olguy** [comment]
Thanks for the contrib @felipeoyarce!

**Comment by felipeoyarce** [comment]
> Thanks for the contrib @felipeoyarce!
> Don't forget to update the changelog with your name and contribution

Oh, thanks @co9olguy :smile:

**Comment by co9olguy** [comment]
@felipeoyarce @thisac Another reminder to add this contribution to the changelog and take credit for your nice contribution :slightly_smiling_face:

**Pull request merged**